### PR TITLE
[docs] Add guide for Undefined Behavior

### PR DESCRIPTION
**Pull request by nunoplopes**
Add a new guide about undefined behavior.

Preview here: https://web.ist.utl.pt/nuno.lopes/UndefinedBehavior.html

This is to be linked from the automatic code review tool, that is now warning about undef (https://github.com/llvm/llvm-project/pull/118506). In the future it could use Alive2 to warn about tests that are full UB.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
I like the idea, and this looks pretty reasonable at a glance.

I think one topic that may be worth mentioning here is the "logical and/or" pattern. It's a non-obvious construction used to get the correct poison propagation semantics.

Not sure if it's out of scope, but we could also mention related APIs, like isGuaranteedNotToBePoison.

**Comment by nunoplopes** [comment]
Thanks for the quick review! I've patched the things you mentioned.

I think having a section about APIs makes sense! For the and/or logic, do we have any API you want to mention? Or do you have an example off the top of your head that we can show?

**Comment by nunoplopes** [comment]
Yes, I agree that we can add a few more things. I'm sure we'll get questions about it, so we can refine as we go.
I just wanted to get something started 🙂  And something to like from the automatic code reviewer. We need more of these automatic things to improve the quality of the code IMHO.

File: llvm/docs/UndefinedBehavior.rst [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+      %ret = add nsw i32 %add, %c
+      ret i32 %ret
+    }
+
+The ``nsw`` attribute in the additions indicates that the operation yields
+poison if there is a signed overflow.
+If the first addition overflows, ``%add`` is poison and thus ``%ret`` is also
+poison since it taints the whole expression DAG.
+
+Poison values can be replaced with any value of type (undef, concrete values,
```
[the remark this answers is not in the document]
Comment (nunoplopes) [reply llvm/docs/UndefinedBehavior.rst:211]: No, you can replace poison with 42. Or replace poison with `freeze poison`. Any replacement of poison is valid, as it is the largest element of the lattice. You can refine it with anything.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by nikic** [comment]
> I think having a section about APIs makes sense! For the and/or logic, do we have any API you want to mention? Or do you have an example off the top of your head that we can show?

For example, we can't transform:
```
define i1 @src(i32 %x, i32 %y) {
    %cmp1 = icmp ne i32 %x, 0
    %cmp2 = icmp ugt i32 %x, %y
    %and = select i1 %cmp1, i1 %cmp2, i1 false
    ret i1 %and
}

define i1 @tgt(i32 %x, i32 %y) {
    %cmp2 = icmp ugt i32 %x, %y
    ret i1 %cmp2
}
```

But we can do it for both of these:

```
define i1 @src2(i32 %x, i32 %y) {
    %cmp1 = icmp ne i32 %x, 0
    %cmp2 = icmp ugt i32 %x, %y
    %and = select i1 %cmp2, i1 %cmp1, i1 false
    ret i1 %and
}

define i1 @src3(i32 %x, i32 %y) {
    %cmp1 = icmp ne i32 %x, 0
    %cmp2 = icmp ugt i32 %x, %y
    %and = and i1 %cmp1, %cmp2
    ret i1 %and
}
```

The logical and/or form has asymmetric poison propagation limiting which transforms are valid.

**Comment by antoniofrighetto** [comment]
@nunoplopes Kindly drop Co-authored-by before merging, as I have not coauthored the change.

File: llvm/docs/UndefinedBehavior.rst [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+      %div = udiv i32 3, %v
+      %r = select i1 %c, i32 %div, i32 0
+      ret i32 %r
+    }
+
+However, this transformation is not correct! Since division triggers UB
+when the divisor is zero, we can only execute speculatively if we are sure we
+don't hit that condition.
+For the function above, when called like ``f(false, 0)``, before the optimization
+it would return 0, and after the optimization it now triggers UB.
```
Comment (antoniofrighetto) [inline llvm/docs/UndefinedBehavior.rst:79]: Nit: readability (up to you).
```suggestion
The function above, when called as ``f(false, 0)``, would return 0 before the
optimization, and triggers UB after being optimizing.
```

**Comment by nunoplopes** [comment]
Thanks @nikic, I'll add that example next.

**Review by shafik — COMMENTED** [review]
Thank you for this write-up. I don't usually sit at this level but I do often work with bug reports dealing w/ UB where this matters.

Maybe it does not fit perfectly here, not sure, but I feel like some greater discussion about some of the more unintuitive effects of UB might be helpful.

One of them that has come up often is when function bodies end up merged due to UB e.g. https://godbolt.org/z/8sEE54E7d

**Comment by nikic** [comment]
@shafik I think this documentation is intended for LLVM developers, not for C/C++ users.

**Pull request merged**